### ISSUE TITLE
Weight Auto-encoder

### ISSUE THREAD

**Issue by csmangum** [csmangum/GCA#11]
Take neural network weights and encode them through another neural network to a N-component transformation (similar to PCA). I also want the model to be able to go from an N-component to full weights (reversing the encoding)

**Comment by csmangum** [comment]
Here's a high-level overview of how you could approach this:

1. **Design the Autoencoder Architecture**: 
   - Your autoencoder will consist of an encoder network and a decoder network.
   - The encoder network will take the full weights of the neural network as input and output the encoded representation (N components).
   - The decoder network will take the encoded representation and reconstruct the full weights.

2. **Training Data**:
   - You'll need a dataset of neural network weights to train your autoencoder.
   - You can generate this dataset by training your neural network on various tasks and then extracting and saving its weights.

3. **Training**:
   - Train the autoencoder using the dataset of neural network weights.
   - The objective will be to minimize the reconstruction error between the original weights and the reconstructed weights.

4. **Encoder and Decoder Implementation**:
   - After training, you can use the encoder part of the autoencoder to encode the full weights into the N component representation.
   - You can then use the decoder part of the autoencoder to decode the N component representation back into the full weights.

5. **Evaluation**:
   - Evaluate the performance of your autoencoder in terms of how accurately it can reconstruct the original weights from the encoded representation.

```python
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np

# Define the Autoencoder class
class Autoencoder(nn.Module):
    def __init__(self, input_dim, encoding_dim):
        super(Autoencoder, self).__init__()
        self.encoder = nn.Sequential(
            nn.Linear(input_dim, 128),
            nn.ReLU(True),
            nn.Linear(128, encoding_dim),
            nn.ReLU(True)
        )
        self.decoder = nn.Sequential(
            nn.Linear(encoding_dim, 128),
            nn.ReLU(True),
            nn.Linear(128, input_dim),
            nn.Sigmoid()
        )

    def forward(self, x):
        x = self.encoder(x)
        x = self.decoder(x)
        return x

# Example usage
input_dim = 1000  # Assuming your original weights have 1000 dimensions
encoding_dim = 50  # Desired number of components

# Instantiate the autoencoder
autoencoder = Autoencoder(input_dim, encoding_dim)

# Define loss function and optimizer
criterion = nn.MSELoss()
optimizer = optim.Adam(autoencoder.parameters(), lr=0.001)

# Generate some dummy data
num_samples = 1000
original_weights = np.random.rand(num_samples, input_dim)
original_weights = torch.tensor(original_weights, dtype=torch.float)

# Training loop
num_epochs = 100
for epoch in range(num_epochs):
    # Forward pass
    outputs = autoencoder(original_weights)
    loss = criterion(outputs, original_weights)
    
    # Backward pass and optimization
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    
    # Print progress
    if (epoch+1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

# Encode and decode
encoded_weights = autoencoder.encoder(original_weights)
reconstructed_weights = autoencoder.decoder(encoded_weights)

# Convert to numpy arrays
encoded_weights = encoded_weights.detach().numpy()
reconstructed_weights = reconstructed_weights.detach().numpy()
```

In this example:

- We define the `Autoencoder` class which inherits from `nn.Module` and includes encoder and decoder layers.
- We use the mean squared error loss and the Adam optimizer for training.
- We generate some dummy data for demonstration purposes.
- We train the autoencoder for a number of epochs and print the loss periodically.
- Finally, we encode and decode the original weights to obtain the reconstructed weights.
